kv/txn_coord_sender.go

```diff
@@ -727,7 +733,7 @@ func (tc *TxnCoordSender) updateState(ctx context.Context, ba roachpb.BatchReque
 				// If the transaction is already over, there's no point in
 				// launching a one-off coordinator which will shut down right
 				// away.
-				if _, isEnding := ba.GetArg(roachpb.EndTransaction); !isEnding {
+				if _, isEnding := ba.GetArg(roachpb.EndTransaction); err == nil && !isEnding {
```

Comment:
Why don't we start the heartbeat loop here? If we're keeping the `txnMeta` open in anticipation of the client's return, don't we need to heartbeat to keep the transaction alive for them?